composer/trainer/trainer.py

```diff
@@ -1653,8 +1655,8 @@ def _train_batch(self, use_grad_scaling: bool):
                             total_loss = self.state.scaler.step(
                                 optimizer, closure=lambda **kwargs: self._train_microbatches(microbatches, **kwargs))
                         else:
-                            total_loss = optimizer.step(
-                                closure=lambda **kwargs: self._train_microbatches(microbatches, **kwargs).item())
+                            total_loss = optimizer.step(closure=lambda **kwargs: self._train_microbatches(
+                                microbatches, **kwargs))  # type: ignore
```

Comment:
Must closures exclusively return floats? Pyright says yes, is it okay to ignore this? The closure in this case will return a dict

Reply:
I think we should probably not be passing around the dictionary to any autograd / optimizer objects, we should just be passing them the total loss Tensor, because that's the only thing that should get `.backward()` called on it, and stuff like that, right? So for logging we pass around the dictionary, but not for any real optimization math, if that makes sense

Reply:
I should read up on closures, but before this PR, `self._train_microbatches` returned `total_loss` which was a detached, cloned copy of the loss, so it didn't have any impact on optimization, right?

Also, if we need the closure to return a single loss value, I could do `self._train_microbatches(microbatches, **kwargs)['total']`, but `total_loss` would not contain the additional losses, making multiple loss logging impossible when using closures. Unless there is a pretty decent refactor of how we pass around `total_loss`

Reply:
Ok gotcha, I think I understand the issue a bit better now... so the thing that worries me a bit is that when optimizers are handed closures, I think they can potentially (e.g. SAM) have some logic inside that checks whether the return value is None, like they expect it to be a float or a None if something went wrong.

I think maybe what we need to do is make the loss stuff an explicit torchmetric (outside the closure) and have the metric.update(...) call happen inside the closure, for each key in the loss dict. Let's brainstorm tomorrow on this if possible!

Reply:
I'll be out Wednesday, let's chat on Thursday if you will be in 🙂

Reply:
Refactored everything, resolving for the next review :)